Generate and refer to source schema type declarations

leemhenson opened the pull request:
## Description

Fixes #3367 

This PR adds an additional step to the build process, before the existing codegen phase occurs. The new step iterates over the source schemas and invokes codegen to produce the minimal step of type definitions for those schemas *untransformed*. These are written into their own files and then imported under an alias into the main codegen output. The alias are used to fully qualify the upstream types. It's easier to demonstrate with an example - it changes this:

```typescript
export type QueryAuthorServiceSdk = {
  author: InContextSdkMethod<Query['author'], QueryauthorArgs, MeshContext>
  // ... other definitions
};
```

to 

```typescript
import * as AuthorService from "./AuthorService";

export type QueryAuthorServiceSdk = {
  author: InContextSdkMethod<AuthorService.Query['author'], AuthorService.QueryAuthorArgs, MeshContext>
  // ... other definitions
};
```

I need to point out that this is a breaking change - the currently released version of mesh is generating sdks based off the upstread schemas *after* applying transformations defined in the `meshrc`. Field renames, filters, etc are all currently visible in the shape of the generated types. This PR changes that choice so codegen emits type definitions against the upstream schema as published. Any users who have code written against those SDKs and have rename transforms (or other shape-altering transforms) applied will get errors from Typescript after regenerating. 

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Screenshots/Sandbox (if appropriate/relevant):

Sandbox links in issue #3367 : https://codesandbox.io/s/condescending-swanson-9n0ux?file=/.mesh/schema.graphql

## How Has This Been Tested?

Tested against a local version of the sandbox.

**Test Environment**:

OS: MacOS 11.6
"@graphql-mesh/cli": "^0.44.4"
"@graphql-mesh/graphql": "^0.20.9"
"@graphql-mesh/transform-filter-schema": "^0.14.5"
NodeJS: 14

## Checklist:

- [x] I have followed the [CONTRIBUTING](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md) doc and the style guidelines of this project
- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests and linter rules pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules

ardatan commented:
Looks good @leemhenson ! Could you add a changeset using `yarn changeset` then we can merge and release?

leemhenson commented:
Hey, I plan to get to this in the next few days. However [per my comment here](https://github.com/Urigo/graphql-mesh/issues/3367#issuecomment-1006947422), the current patch is insufficient in the presence of field renames. I am thinking about collecting a map of renamed field to original field name and using that during the codegen process to generate the correct type names.

leemhenson commented on packages/cli/src/commands/ts-artifacts.ts:
```diff
@@ -151,15 +197,17 @@ export async function generateTsArtifacts({
               const item = generateTypesForApi({
-                schema: sourceSchema,
-                name: source.name,
+                schema: source.schema,
```
This is where I've switched the context sdks to be built against the untransformed upstream schemas.

ardatan replied:
InContext SDK calls `delegateToSchema` of GraphQL Tools against transformed schema. In this PR's scope, we need to follow it. We can discuss having different SDKs for other use cases in a different discussion or PRs later :)

leemhenson commented:
@ardatan This PR now represents a breaking change, and as such would necessitate a change from `0.x` to `1.x`. `yarn changest` suggests contacting the package owner to verify that is intended, so here we are. 😄

leemhenson commented:
@ardatan I've reverted the change I had made so it's back to using the transformed schema as the source. I've also now added a map to `WrapRename` which is populated as the transformations are applied. The artifact generation can then look for `WrapRename` instances and use them during the codegen to figure out what the original field name was.

ardatan commented on packages/transforms/rename/test/wrapRename.spec.ts:
```diff
@@ -481,4 +481,41 @@ describe('rename', () => {
     expect(printSchema(newSchema)).toMatchSnapshot();
   });
+
+  it('should track the original name of a renamed field', () => {
```
Why do we track renamed fields? In context SDK is passing requests to the transformed schema already.

leemhenson replied:
In order to get back to the original field name, which must be used in order to generate valid type declarations: https://github.com/Urigo/graphql-mesh/pull/3414/files#diff-8d4bc1ae7b41e598b72810975c4e0722c93268daa0b53000d30ebf3d5c4d4886R51

ardatan replied:
If we generate types from transformedSchema, we shouldn't need to track these. Maybe I'm missing something here.

ardatan replied:
Oh you're trying to implement another SDK. But can we please seperate it into two different PRs?
- Generate individual types for Incontext SDK with namespaces for each transformed source
- Have another Incontext SDK for the original schema.

leemhenson replied:
No I'm not implementing another SDK. This is the same SDK, but in the presence of rename transforms it fails to generate the typescript declarations because the field names do not match. This PR emits source schema type declarations and tracks the renames so that the generate declarations refer to the correct source schema types.

leemhenson replied:
To clarify (as original reported in issue #3367 ): if you rename fields in the source schema, and then filter those fields out in the root filter, the `.mesh/index.d.ts` type declarations are not emitted. The sandbox that demonstrates it is here: https://codesandbox.io/s/condescending-swanson-9n0ux?file=/.mesh/schema.graphql

Just run `yarn build` and observe that the declarations are not emitted. If you remove the root filter, they are. If you modify the artifact generation so it writes the "intermediate" TS to disk, when you look inside you'll see the field name mismatch.

leemhenson commented:
Bit stumped on this one now. It all looks fine in terms of the emitted types, but when I attempt to run the mesh server it explodes in my generated `.mesh/index.mjs` because the `RenameTransform` in `new RenameTransform` at runtime is not a class instance - it's instead an object that looks like:

`{ default: <function>, WrapRename: <class> }` 

I assume this is because I've added the named export to expose the `WrapRename` class so I can import it into the artifacts generation. Is there some config I can change to make this work?

The other approach I have tried is not adding the named export for `WrapRename`, but instead directly importing it with `import WrapRename from "@graphql-mesh/transform-rename/wrapRename"`. This seems to work if I add the path to tsconfig.json, but at run time it can't find the js file: 

```
Error: Cannot find module '<my project path>/node_modules/@graphql-mesh/transform-rename/wrapRename.js'
```

because the mesh build process puts everything into index.js. Is this solvable?

leemhenson commented:
@ardatan is there anything else I can do here? I can generate a changeset but I'm not sure what perspective you would have on it. I'm changing the emitted code, so does that constitute a breaking change for the `cli` package?

ardatan commented:
Based on your feedback and PR, we have implemented it. Now we generate types individually for each source. Could you check if that solves your issue?

leemhenson commented:
latest release seems to work ok for me 👍
